### PR TITLE
Fix Regex features in raw-string-literals.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/65836

### CODE REVIEW

File: src/EditorFeatures/Test2/ReferenceHighlighting/CSharpReferenceHighlightingTests.vb [hunk 1]
```diff
@@ -725,8 +695,57 @@ class C
...
+               <Project Language="C#" CommonReferences="true">
+                   <Document>
+using System.Text.RegularExpressions;
+
+class C
+{
+    void Goo()
+    {
+        var r = new Regex("""
+    $$       (a)\0\1
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/ReferenceHighlighting/CSharpReferenceHighlightingTests.vb:737]: this previously would crash in highlight refs.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests_Regex.vb [hunk 2]
```diff
@@ -51,6 +72,74 @@ class c
...
+        <WpfTheory, CombinatorialData>
+        Public Async Function ExplicitInvoke_RawSingleLineString(showCompletionInArgumentLists As Boolean) As Task
+            Using state = TestStateFactory.CreateCSharpTestState(
+                <Document><![CDATA[
+using System.Text.RegularExpressions;
+class c
+{
+    void goo()
+    {
+        var r = new Regex("""$$ """);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests_Regex.vb:106]: this case would previously assert, but not crash.

File: src/Features/CSharp/Portable/EmbeddedLanguages/EmbeddedLanguageUtilities.cs [hunk 3]
```diff
@@ -18,21 +17,22 @@ internal static void AddComment(SyntaxEditor editor, SyntaxToken stringLiteral,
...
 
         public static string EscapeText(string text, SyntaxToken token)
         {
-            // This function is called when Completion needs to escape something its going to
-            // insert into the user's string token.  This means that we only have to escape
-            // things that completion could insert.  In this case, the only regex character
-            // that is relevant is the \ character, and it's only relevant if we insert into
-            // a normal string and not a verbatim string.  There are no other regex characters
-            // that completion will produce that need any escaping. 
-            Debug.Assert(token.Kind() == SyntaxKind.StringLiteralToken);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/EmbeddedLanguages/EmbeddedLanguageUtilities.cs:32]: bogus assert.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualCharSequence.Chunks.cs [hunk 4]
```diff
@@ -62,7 +62,12 @@ public ImmutableSegmentedListChunk(ImmutableSegmentedList<VirtualChar> array)
 
                     return 0;
                 });
-                Debug.Assert(index >= 0);
+
+                // Characters can be discontiguous (for example, in multi-line-raw-string literals).  So if the
+                // position is in one of the gaps, it won't be able to find a corresponding virtual char.
+                if (index < 0)
+                    return null;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/VirtualCharSequence.Chunks.cs:69]: actual crash fix.

**Review by akhera99 — APPROVED** [review]

**Pull request merged**